MaterialLib/Fluid/Density/LiquidDensity.h

```diff
@@ -0,0 +1,159 @@
+    _p0(config.getConfigParameter<double>("p0")),
+          //! \ogs_file_param{material__fluid__density__liquid_density__bulk_modulus}
+    _bulk_moudlus(config.getConfigParameter<double>("bulk_modulus"))
+    {
+    }
+
+    /// Get density model name.
+    std::string getName() const override
+    {
+        return "Temperature or pressure dependent liquid density";
+    }
+
+    /// Get density value.
+    /// \param var_vals Variable values in an array. The order of its elements
+    ///                 is given in enum class PropertyVariable.
+    double getValue(const double var_vals[]) const override
+    {
+        const double T = var_vals[static_cast<int>(PropertyVariable::T)];
+        const double p = var_vals[static_cast<int>(PropertyVariable::pl)];
+        return _rho0 / (1. + _beta * (T - _temperature0)) /
+               (1. - (p - _p0) / _bulk_moudlus);
+    }
+
+    /// Get the partial differential of the density with respect to temperature
+    /// or liquid pressure.
+    /// \param var_vals  Variable values  in an array. The order of its elements
+    ///                   is given in enum class PropertyVariable.
+    /// \param var       Variable type.
+    double getdValue(const double var_vals[],
+            const PropertyVariable var) const override
+    {
+        assert(var == PropertyVariable::T || var == PropertyVariable::pl);
+
+        const int func_id = (var == PropertyVariable::T) ? 0 : 1;
+        const double T = var_vals[static_cast<int>(PropertyVariable::T)];
+        const double p = var_vals[static_cast<int>(PropertyVariable::pl)];
+        return (this->*_derivative_functions[func_id])(T, p);
+    }
+
+private:
+    /// Volumetric temperature expansion coefficient.
+    const double _beta;
+
+    /// Initial density.
+    const double _rho0;
+
+    /// Initial temperature.
+    const double _temperature0;
+
+    /// Initial pressure.
+    const double _p0;
+
+    /// Bulk modulus.
+    const double _bulk_moudlus;
+
+    /**
+     *  Calculate the derivative of density of fluids with respect to
+     * temperature.
+     *   \param T    Temperature (K).
+     *   \param p    Pressure (Pa).
+     */
+    double dLiquidDensity_dT(const double T, const double p) const
+    {
+        const double fac_T = 1. + _beta * (T - _temperature0);
+        return -_beta * _rho0 / (fac_T * fac_T) /
+               (1. - (p - _p0) / _bulk_moudlus);
+    }
+
+    /**
+     *  Calculate the derivative of density of fluids with respect to pressure.
+     *   \param T    Temperature (K).
+     *   \param p    Pressure (Pa).
+     */
+    double dLiquidDensity_dp(const double T, const double p) const
+    {
+        const double fac_p = 1. - (p - _p0) / _bulk_moudlus;
+        return _rho0 / (1. + _beta * (T - _temperature0)) /
+               (fac_p * fac_p * _bulk_moudlus);
+    }
+
+    typedef double (LiquidDensity::*DerivativeFunctionPointer)(const double,
```

Comment:
After this line, there is a declaration of function pointer array
```
 static DerivativeFunctionPointer _derivative_functions[2];
```
The array is used in the member function of getdValue, which calculates the derivative with respect to a variable.